File: services/engagement/src/engagement/analytics/peer_percentile.py
"""Sprint 32 (P4-S32) — pure-function peer-percentile helpers.

Per-(user, topic, exam) percentile rank vs the platform cohort, with an
anonymity threshold (NFR-P4-06) that hides the result when the cohort is
small.

Pure: no DB / HTTP coupling.
"""

from __future__ import annotations

from typing import Any

DEFAULT_ANONYMITY_THRESHOLD = 30


def compute_peer_percentile(
    user_ewa: float, peer_ewas: list[float]
) -> float | None:
    """Strict-below percentile: fraction of peers with EWA < user_ewa.

    Returns 0.0 when user is at or below every peer (or peers is empty
    via the surrogate); None when the peer list is genuinely empty so
    callers can decide to hide rather than render "0%".
    """
    if not peer_ewas:
        return None
    below = sum(1 for e in peer_ewas if float(e) < float(user_ewa))
    return round(100.0 * below / len(peer_ewas), 2)


def is_anonymity_threshold_met(
    peer_count: int, *, threshold: int = DEFAULT_ANONYMITY_THRESHOLD
) -> bool:
    """Hide the percentile when the cohort is small (NFR-P4-06)."""
    return peer_count >= threshold


def summarise_percentile(
    user_ewa: float,
    peer_ewas: list[float],
    *,
    threshold: int = DEFAULT_ANONYMITY_THRESHOLD,
) -> dict[str, Any]:
    """Return either a hidden marker (with reason) or the visible
    percentile + cohortSize. Driver for the route + UI.
    """
    cohort = len(peer_ewas)
    if not is_anonymity_threshold_met(cohort, threshold=threshold):
        return {
            "hidden": True,
            "reason": "cohort_too_small",
            "cohortSize": cohort,
            "thresholdRequired": threshold,
        }
    pct = compute_peer_percentile(user_ewa, peer_ewas)
    return {
        "hidden": False,
        "percentile": pct,
        "cohortSize": cohort,
        "userEwa": round(float(user_ewa), 4),
    }

File: apps/web-student/src/lib/peer_percentile.ts
// Sprint 32 (P4-S32) — pure helpers for the per-topic peer-percentile pill.
//
// The endpoint hides the result when cohort < 30 (NFR-P4-06). The pill
// renders three states: hidden / visible-with-band / cold-start.

export interface PeerPercentileResp {
  userId: string;
  examId: string;
  topicId: string;
  hidden: boolean;
  reason?: string;
  cohortSize: number;
  thresholdRequired?: number;
  percentile?: number;
  userEwa?: number;
}

export type PercentilePillBand = "top" | "high" | "mid" | "low";

export interface PercentilePillState {
  visible: boolean;
  /** When hidden, why — drives a small disclosure to the user. */
  hideReason?: string;
  /** Visible-state-only fields */
  label?: string; // "67th percentile (N=230)"
  band?: PercentilePillBand;
}

function ordinalSuffix(n: number): string {
  const s = ["th", "st", "nd", "rd"];
  const v = Math.round(n) % 100;
  return s[(v - 20) % 10] || s[v] || s[0];
}

export function bandFor(percentile: number): PercentilePillBand {
  if (percentile >= 90) return "top";
  if (percentile >= 70) return "high";
  if (percentile >= 30) return "mid";
  return "low";
}

export function pillState(resp: PeerPercentileResp | null): PercentilePillState {
  if (resp === null) return { visible: false };
  if (resp.hidden) {
    return { visible: false, hideReason: resp.reason };
  }
  if (resp.percentile === undefined) return { visible: false };
  const pct = Math.round(resp.percentile);
  return {
    visible: true,
    label: `${pct}${ordinalSuffix(pct)} percentile (N=${resp.cohortSize})`,
    band: bandFor(resp.percentile),
  };
}
